In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
PATH = 'data/'

In [3]:
products = pd.read_csv(f'{PATH}amazon_baby.csv')

In [4]:
products = products.fillna({'review':''})

In [5]:
def remove_punctuation(text):
    import string
    return text.translate(str.maketrans('', '', string.punctuation)) 

products['review_clean'] = products['review'].apply(remove_punctuation)

In [6]:
products = products[products['rating'] != 3]

In [7]:
products['sentiment'] = products['rating'].apply(lambda rating: +1 if rating > 3 else -1)

In [13]:
import json
f = open(f'{PATH}module-2-assignment-train-idx.json',)
train_ids = json.load(f)
f = open(f'{PATH}module-2-assignment-test-idx.json',)
test_ids = json.load(f)
f.close()

In [14]:
train_data = products.iloc[train_ids]

In [15]:
test_data = products.iloc[test_ids]

In [16]:
train_data

name  \
1                                   Planetwise Wipe Pouch   
2                     Annas Dream Full Quilt with 2 Shams   
3       Stop Pacifier Sucking without tears with Thumb...   
4       Stop Pacifier Sucking without tears with Thumb...   
5       Stop Pacifier Sucking without tears with Thumb...   
...                                                   ...   
183525  Squeasy Snacker 6oz Silicone Reusable Food Pou...   
183526  Baby Teething Necklace for Mom Pretty Donut Sh...   
183527  Baby Teething Necklace for Mom Pretty Donut Sh...   
183528   Abstract 2 PK Baby / Toddler Training Cup (Pink)   
183529  Baby Food Freezer Tray - Bacteria Resistant, B...   

                                                   review  rating  \
1       it came early and was not disappointed. i love...       5   
2       Very soft and comfortable and warmer than it l...       5   
3       This is a product well worth the purchase.  I ...       5   
4       All of my kids have cried non-stop when I trie...       5   
5       When the Binky Fairy came to our house, we did...       5   
...                                                   ...     ...   
183525  This reusable squeeze bottle is the best I hav...       5   
183526  Such a great idea! very handy to have and look...       5   
183527  This product rocks!  It is a great blend of fu...       5   
183528  This item looks great and cool for my kids.......       5   
183529  I am extremely happy with this product. I have...       5   

                                             review_clean  sentiment  
1       it came early and was not disappointed i love ...          1  
2       Very soft and comfortable and warmer than it l...          1  
3       This is a product well worth the purchase  I h...          1  
4       All of my kids have cried nonstop when I tried...          1  
5       When the Binky Fairy came to our house we didn...          1  
...                                                   ...        ...  
183525  This reusable squeeze bottle is the best I hav...          1  
183526  Such a great idea very handy to have and look ...          1  
183527  This product rocks  It is a great blend of fun...          1  
183528  This item looks great and cool for my kidsI kn...          1  
183529  I am extremely happy with this product I have ...          1  

[133416 rows x 5 columns]

In [17]:
test_data

name  \
9       Baby Tracker&reg; - Daily Childcare Journal, S...   
10      Baby Tracker&reg; - Daily Childcare Journal, S...   
16         Nature's Lullabies First Year Sticker Calendar   
20        Nature's Lullabies Second Year Sticker Calendar   
28                            Lamaze Peekaboo, I Love You   
...                                                   ...   
183507  Maxboost iPhone 5S/5 Case - Protective Snap-on...   
183515  Maxboost iPhone 5S/5 Case - Protective Snap-on...   
183522  Airline Seat Belt Extender - The Best Extensio...   
183524  Squeasy Snacker 6oz Silicone Reusable Food Pou...   
183530  Best 2 Pack Baby Car Shade for Kids - Window S...   

                                                   review  rating  \
9       This has been an easy way for my nanny to reco...       4   
10      I love this journal and our nanny uses it ever...       4   
16      I love this little calender, you can keep trac...       5   
20      I had a hard time finding a second year calend...       5   
28      One of baby's first and favorite books, and it...       4   
...                                                   ...     ...   
183507  got this for my wife and she loves it would de...       5   
183515  I love this phone case!  My  iPhone is always ...       5   
183522  I bought this as a father's day gift for my da...       5   
183524  I love this product, it makes my life easier. ...       5   
183530  I love this product very mush . I have bought ...       5   

                                             review_clean  sentiment  
9       This has been an easy way for my nanny to reco...          1  
10      I love this journal and our nanny uses it ever...          1  
16      I love this little calender you can keep track...          1  
20      I had a hard time finding a second year calend...          1  
28      One of babys first and favorite books and it i...          1  
...                                                   ...        ...  
183507  got this for my wife and she loves it would de...          1  
183515  I love this phone case  My  iPhone is always s...          1  
183522  I bought this as a fathers day gift for my dad...          1  
183524  I love this product it makes my life easier Wi...          1  
183530  I love this product very mush  I have bought m...          1  

[33336 rows x 5 columns]

In [18]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer(token_pattern=r'\b\w+\b')
     # Use this token pattern to keep single-letter words
# First, learn vocabulary from the training data and assign columns to words
# Then convert the training data into a sparse matrix
train_matrix = vectorizer.fit_transform(train_data['review_clean'])
# Second, convert the test data into a sparse matrix, using the same word-column mapping
test_matrix = vectorizer.transform(test_data['review_clean'])

In [19]:
from sklearn.linear_model import LogisticRegression

In [22]:
sentiment_model = LogisticRegression(max_iter=10000).fit(train_matrix, train_data['sentiment'])

In [26]:
sentiment_model.coef_.size

121712

### Sentiment model has 121712 coefficients

In [31]:
(sentiment_model.coef_ >= 0).sum()

91020

### 91020 coefficients are non zero

In [36]:
sample_test_data = test_data.iloc[10:13]
print(sample_test_data)

                                                 name  \
59                          Our Baby Girl Memory Book   
71  Wall Decor Removable Decal Sticker - Colorful ...   
91  New Style Trailing Cherry Blossom Tree Decal R...   

                                               review  rating  \
59  Absolutely love it and all of the Scripture in...       5   
71  Would not purchase again or recommend. The dec...       2   
91  Was so excited to get this product for my baby...       1   

                                         review_clean  sentiment  
59  Absolutely love it and all of the Scripture in...          1  
71  Would not purchase again or recommend The deca...         -1  
91  Was so excited to get this product for my baby...         -1  


In [37]:
sample_test_data.iloc[0]['review']

'Absolutely love it and all of the Scripture in it.  I purchased the Baby Boy version for my grandson when he was born and my daughter-in-law was thrilled to receive the same book again.'

In [38]:
sample_test_data.iloc[1]['review']

'Would not purchase again or recommend. The decals were thick almost plastic like and were coming off the wall as I was applying them! The would NOT stick! Literally stayed stuck for about 5 minutes then started peeling off.'

In [39]:
sample_test_matrix = vectorizer.transform(sample_test_data['review_clean'])
scores = sentiment_model.decision_function(sample_test_matrix)
print(scores)

[  5.59998522  -3.16970521 -10.42292433]


In [40]:
sentiment_model.predict(sample_test_matrix)

array([ 1, -1, -1], dtype=int64)

In [41]:
def sigmoid(x): return 1/(1+np.exp(-x))

In [43]:
for score in scores:
    print(sigmoid(score))

0.9963157058323063
0.040321820978886216
2.974188988613961e-05


In [44]:
sentiment_model.predict_proba(sample_test_matrix)

array([[3.68429417e-03, 9.96315706e-01],
       [9.59678179e-01, 4.03218210e-02],
       [9.99970258e-01, 2.97418899e-05]])

### 3rd data point in sample test data has the lowest probability of being declared as a positive review

In [46]:
test_scores = sentiment_model.decision_function(test_matrix)

In [56]:
best_20_indices = test_scores.argsort()[-20:]

In [52]:
worst_20_indices = test_scores.argsort()[:20]

In [58]:
worst_20_indices

array([ 2931, 21700, 13939,  8818, 28184, 17069,  9655, 14711,  1942,
       20594, 10814,  1810,  7310, 31226, 13751, 27231, 28120,   205,
        5831, 15062], dtype=int64)

### The following are the best 20 ratings

In [57]:
test_data.iloc[best_20_indices]

name  \
147996  Baby Jogger City Mini GT Double Stroller, Shad...   
182089  Summer Infant Wide View Digital Color Video Mo...   
165593  Ikea 36 Pcs Kalas Kids Plastic BPA Free Flatwa...   
22586      Britax Decathlon Convertible Car Seat, Tiffany   
66059        Evenflo 6 Pack Classic Glass Bottle, 4-Ounce   
114796  Fisher-Price Cradle 'N Swing,  My Little Snuga...   
147949  Baby Jogger City Mini GT Single Stroller, Shad...   
97325   Freemie Hands-Free Concealable Breast Pump Col...   
168697  Graco FastAction Fold Jogger Click Connect Str...   
52631   Evenflo X Sport Plus Convenience Stroller - Ch...   
80155   Simple Wishes Hands-Free Breastpump Bra, Pink,...   
168081  Buttons Cloth Diaper Cover - One Size - 8 Colo...   
180646      Mamas &amp; Papas 2014 Urbo2 Stroller - Black   
119182  Roan Rocco Classic Pram Stroller 2-in-1 with B...   
50315          P'Kolino Silly Soft Seating in Tias, Green   
137034         Graco Pack 'n Play Element Playard - Flint   
140816         Diono RadianRXT Convertible Car Seat, Plum   
133651                  Britax 2012 B-Agile Stroller, Red   
87017     Baby Einstein Around The World Discovery Center   
100166  Infantino Wrap and Tie Baby Carrier, Black Blu...   

                                                   review  rating  \
147996  We are well pleased with this stroller, and I ...       4   
182089  I love this baby monitor.  I can compare this ...       5   
165593  For the price this set is unbelievable- and tr...       5   
22586   I researched a few different seats to put in o...       4   
66059   It's always fun to write a review on those pro...       5   
114796  My husband and I cannot state enough how much ...       5   
147949  Amazing, Love, Love, Love it !!! All 5 STARS a...       5   
97325   I absolutely love this product.  I work as a C...       5   
168697  Graco's FastAction Jogging Stroller definitely...       5   
52631   After seeing this in Parent's Magazine and rea...       5   
80155   I just tried this hands free breastpump bra, a...       5   
168081  We are big Best Bottoms fans here, but I wante...       4   
180646  After much research I purchased an Urbo2. It's...       4   
119182  Great Pram Rocco!!!!!!I bought this pram from ...       5   
50315   I've purchased both the P'Kolino Little Reader...       4   
137034  My husband and I assembled this Pack n' Play l...       4   
140816  I bought this seat for my tall (38in) and thin...       5   
133651  [I got this stroller for my daughter prior to ...       4   
87017   I am so HAPPY I brought this item for my 7 mon...       5   
100166  I bought this carrier when my daughter was abo...       5   

                                             review_clean  sentiment  
147996  We are well pleased with this stroller and I w...          1  
182089  I love this baby monitor  I can compare this o...          1  
165593  For the price this set is unbelievable and tru...          1  
22586   I researched a few different seats to put in o...          1  
66059   Its always fun to write a review on those prod...          1  
114796  My husband and I cannot state enough how much ...          1  
147949  Amazing Love Love Love it  All 5 STARS all the...          1  
97325   I absolutely love this product  I work as a Cu...          1  
168697  Gracos FastAction Jogging Stroller definitely ...          1  
52631   After seeing this in Parents Magazine and read...          1  
80155   I just tried this hands free breastpump bra an...          1  
168081  We are big Best Bottoms fans here but I wanted...          1  
180646  After much research I purchased an Urbo2 Its e...          1  
119182  Great Pram RoccoI bought this pram from Europe...          1  
50315   Ive purchased both the PKolino Little Reader C...          1  
137034  My husband and I assembled this Pack n Play la...          1  
140816  I bought this seat for my tall 38in and thin 2...          1  
133651  I got this stroller for my daughter

### The following are the worst 20 ratings

In [59]:
test_data.iloc[worst_20_indices]

name  \
16042         Fisher-Price Ocean Wonders Aquarium Bouncer   
120209  Levana Safe N'See Digital Video Baby Monitor w...   
77072      Safety 1st Exchangeable Tip 3 in 1 Thermometer   
48694   Adiri BPA Free Natural Nurser Ultimate Bottle ...   
155287  VTech Communications Safe &amp; Sounds Full Co...   
94560   The First Years True Choice P400 Premium Digit...   
53207                 Safety 1st High-Def Digital Monitor   
81332               Cloth Diaper Sprayer--styles may vary   
10677                   Philips AVENT Newborn Starter Set   
113995  Motorola Digital Video Baby Monitor with Room ...   
59546              Ellaroo Mei Tai Baby Carrier - Hershey   
9915         Cosco Alpha Omega Elite Convertible Car Seat   
40079   Chicco Cortina KeyFit 30 Travel System in Adve...   
172090  Belkin WeMo Wi-Fi Baby Monitor for Apple iPhon...   
75994          Peg-Perego Tatamia High Chair, White Latte   
149987                   NUK Cook-n-Blend Baby Food Maker   
154878  VTech Communications Safe &amp; Sound Digital ...   
1116                Safety 1st Deluxe 4-in-1 Bath Station   
31741              Regalo My Cot Portable Bed, Royal Blue   
83234       Thirsties Hemp Inserts 2 Pack, Small 6-18 Lbs   

                                                   review  rating  \
16042   We have not had ANY luck with Fisher-Price pro...       2   
120209  This is the first review I have ever written o...       1   
77072   I thought it sounded great to have different t...       1   
48694   I will try to write an objective review of the...       2   
155287  This is my second video monitoring system, the...       1   
94560   Note: we never installed batteries in these un...       1   
53207   We bought this baby monitor to replace a diffe...       1   
81332   I bought this sprayer out of desperation durin...       1   
10677   It's 3am in the morning and needless to say, t...       1   
113995  DO NOT BUY THIS BABY MONITOR!I purchased this ...       1   
59546   This is basically an overpriced piece of fabri...       1   
9915    I bought this car seat after both seeing  the ...       1   
40079   My wife and I have used this system in two car...       1   
172090  I read so many reviews saying the Belkin WiFi ...       2   
75994   I can see why there are so many good reviews o...       2   
149987  It thought this would be great. I did a lot of...       1   
154878  First, the distance on these are no more than ...       1   
1116    This item is junk.  I originally chose it beca...       1   
31741   If I could give this product zero stars I woul...       1   
83234   My Experience: Babykicks Inserts failure vs RA...       5   

                                             review_clean  sentiment  
16042   We have not had ANY luck with FisherPrice prod...         -1  
120209  This is the first review I have ever written o...         -1  
77072   I thought it sounded great to have different t...         -1  
48694   I will try to write an objective review of the...         -1  
155287  This is my second video monitoring system the ...         -1  
94560   Note we never installed batteries in these uni...         -1  
53207   We bought this baby monitor to replace a diffe...         -1  
81332   I bought this sprayer out of desperation durin...         -1  
10677   Its 3am in the morning and needless to say thi...         -1  
113995  DO NOT BUY THIS BABY MONITORI purchased this m...         -1  
59546   This is basically an overpriced piece of fabri...         -1  
9915    I bought this car seat after both seeing  the ...         -1  
40079   My wife and I have used this system in two car...         -1  
172090  I read so many reviews saying the Belkin WiFi ...         -1  
75994   I can see why there are so many good reviews o...         -1  
149987  It thought this would be great I did a lot of ...         -1  
154878  First the distance on these are no more than 7...         -1  
1116    This item is junk  I originally cho

In [60]:
test_predictions = sentiment_model.predict(test_matrix)

In [62]:
test_ground_labels = test_data['sentiment']

In [65]:
(test_predictions == test_ground_labels).sum() / len(test_predictions)

0.9322954163666907

### Accuracy of sentiment model on test data in 93.23%

### Higher accuracy on training data doesn't always imply a better classifier

In [66]:
significant_words = ['love', 'great', 'easy', 'old', 'little', 'perfect', 'loves', 
      'well', 'able', 'car', 'broke', 'less', 'even', 'waste', 'disappointed', 
      'work', 'product', 'money', 'would', 'return']

In [67]:
vectorizer_word_subset = CountVectorizer(vocabulary=significant_words) # limit to 20 words
train_matrix_word_subset = vectorizer_word_subset.fit_transform(train_data['review_clean'])
test_matrix_word_subset = vectorizer_word_subset.transform(test_data['review_clean'])

In [68]:
simple_model = LogisticRegression().fit(train_matrix_word_subset,train_data['sentiment'])

In [69]:
simple_model_coef_table = pd.DataFrame({'word':significant_words,
                                         'coefficient':simple_model.coef_.flatten()})

In [71]:
simple_model_coef_table.sort_values(by='coefficient',ascending=False)

word  coefficient
6          loves     1.673269
5        perfect     1.510263
0           love     1.363697
2           easy     1.192219
1          great     0.943950
4         little     0.520174
7           well     0.503760
8           able     0.190937
3            old     0.085424
9            car     0.058813
11          less    -0.209348
16       product    -0.320491
18         would    -0.362157
12          even    -0.511456
15          work    -0.621307
17         money    -0.898062
10         broke    -1.652144
13         waste    -2.034489
19        return    -2.109815
14  disappointed    -2.348478

### 10 coefficients are positive for the simple model

In [79]:
sentiment_model_coef_table = pd.DataFrame({'word':list(vectorizer.vocabulary_.keys()),
                                         'coefficient':sentiment_model.coef_.flatten()})

In [80]:
sentiment_model_coef_table

word  coefficient
0                 it    -1.239576
1               came     0.000045
2              early     0.027483
3                and     0.003578
4                was     0.000017
...              ...          ...
121707  uprightright     0.006564
121708   squeezeable     0.000008
121709       squeasy     0.012532
121710       unllike     0.002702
121711         bimbi    -0.000043

[121712 rows x 2 columns]

In [87]:
for word in simple_model_coef_table.sort_values(by='coefficient',ascending=False)['word']:
    print(sentiment_model_coef_table[sentiment_model_coef_table['word'] == word])

     word  coefficient
74  loves     0.010474
        word  coefficient
181  perfect    -0.686908
   word  coefficient
8  love     0.267401
    word  coefficient
98  easy    -0.005556
      word  coefficient
114  great     0.065302
      word  coefficient
76  little    -0.314856
    word  coefficient
46  well     0.000005
     word  coefficient
164  able     0.212654
    word  coefficient
462  old     0.008179
    word  coefficient
473  car     0.055098
      word  coefficient
1997  less     0.041779
       word  coefficient
45  product     0.023209
      word  coefficient
150  would     0.219904
     word  coefficient
682  even     0.079141
     word  coefficient
100  work     0.000129
      word  coefficient
491  money     0.000585
       word  coefficient
2574  broke    -0.720655
      word  coefficient
642  waste     0.005789
        word  coefficient
2843  return     0.198064
           word  coefficient
6  disappointed     0.002829


### Positive words in the simple model are not the same as positive words in the sentiment model

In [88]:
train_predictions = sentiment_model.predict(train_matrix)
train_ground_labels = train_data['sentiment']
(train_predictions == train_ground_labels).sum() / len(train_predictions)

0.9684895364873778

### Accuracy of sentiment model on train data in 96.85%

In [89]:
train_predictions = simple_model.predict(train_matrix_word_subset)
train_ground_labels = train_data['sentiment']
(train_predictions == train_ground_labels).sum() / len(train_predictions)

0.8668225700065959

### Accuracy of simple model on train data in 86.68%

### Sentiment model has higher accuracy on training set

In [90]:
test_predictions = simple_model.predict(test_matrix_word_subset)
test_ground_labels = test_data['sentiment']
(test_predictions == test_ground_labels).sum() / len(test_predictions)

0.8693604511639069

### Accuracy of simple model on test data is 86.94%

### Sentiment model has higher accuracy on test set

In [93]:
len(test_data[test_data['sentiment'] == 1])/len(test_data)

0.8427825773938085

### Accuracy of majority class classifier on test set is 84.28%

### The sentiment model is definitely better than the majority class classifier